ENH: Added sum normalization to histogram()

the-rccg opened the pull request:
With matplotlib and pandas using histogram() for charting, this allows for fractional histograms by adding sum=1. It also removes that passive aggressive note about it not doing sum normalization by adding a one liner so people stop having to read that comment when looking for the functionality. 
Since strings default validate as True, nothing changes for existing codes and typos.

Commit:
implemented percentage histogram values

new parameter value, density==2: n/n.sum() in order to allow for fractional/percentage histogram charting in matplotlib and pandas

Commit:
added density==2 explanation

Commit:
changed sum requirement from 2 -> 'sum'

eric-wieser commented:
Can you expand on the use case for a "fractional histogram"?

To me it seems that this is better written explicitly as:
```
h, edges = np.histogram(...)
h /= h.sum()  # scale so counts (not areas) sum to 1
```

the-rccg commented:
Generally speaking, given an array, this allows directly to show what fraction of the array is between certain ranges/within the bin. 

Since a density parameter is already implemented and the potential usefulness has been acknowledged by saying the parameter does not do a sum normalization in the function parameter description, it is a minor addition to add this functionality for downstream packages. Specifically, the Matplotlib Histogram plots that are then used again in Pandas histogram plots. While both allow to pass the density parameter through to np.histogram(), they draw the shapes before you can normalize them by the sum, and you are left with having to loop over and change the height of each shape individually, Line 6271 in the Axes class [here](https://github.com/matplotlib/matplotlib/blob/0f6d9f9704e782901aefc5f3594b3e87e04c0d04/lib/matplotlib/axes/_axes.py#L6271). 
There is an argument to be made to change this in Matplotlib rather than NumPy, but for consistency of determining bin height and edges, it makes the most sense to add this functionality to where they are calculated, which already has density normalization implemented setting the precedence for implementing the sum normalization here. 

Bottom line, given the way the function is used in major packages, and given related functionality with the parameter is already implemented, it makes most sense to add it here for logical consistency.

Commit:
Clarified that 'sum' is a PMF

Changed "this is *not* a probability mass function" of integral normalized parameter, to now read "this *is* a probability mass function" for the added sum normalization functionality. Potentially allowing 'pmf' and 'sum' would make sense.

eric-wieser commented on numpy/lib/histograms.py:
```diff
@@ -605,13 +605,14 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
+        the *integral* over the range is 1. 
+        If ``sum``, the result will be normalized such that the sum of 
+        the histogram values will be equal to 1. This *is* a probability 
+        mass function. 
```
It seems very peculiar to me to use the `density` argument to request a `mass` function. I'd prefer a `mass=True` argument to acheive that, or perhaps `mode='count'/'density'/'mass'`.

I think there have been discussions about this before, either here or on the mailing list.

eric-wieser commented:
> they draw the shapes before you can normalize them by the sum

You can fix this with:
```
plt.hist(data, bins, weights=1/len(data))
```

eric-wieser commented:
Closing this, since:
* The implementation is undesirable
* We should keep discussion about whether we want this at all to the original issue
